core/tests/unit/ldml/kmx_plus.tests.cpp

```diff
@@ -163,6 +163,162 @@ TEST(KMXPlusTest, test_uset) {
+    EXPECT_TRUE(COMP_KMXPLUS_STRS::valid_string(s, 5)); // whole thing minus 'c'
+    EXPECT_TRUE(COMP_KMXPLUS_STRS::valid_string(s+2, 3)); // just the marker
+    // if we slice the marker, invalid
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 4));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 3));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s+2, 2));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s+2, 1));
+  }
+  {
+    // valid str with a marker - 'ab\m{x}c'
+    km_core_cu const s[] = {u'a', u'b', 0xFFFF, 0x0008, 0x0001, u'c'};
+    EXPECT_TRUE(COMP_KMXPLUS_STRS::valid_string(s, 6));      // whole thing
+    EXPECT_TRUE(COMP_KMXPLUS_STRS::valid_string(s, 5));      // whole thing minus 'c'
+    EXPECT_TRUE(COMP_KMXPLUS_STRS::valid_string(s + 2, 3));  // just the marker
+    // if we slice the marker, invalid
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 4));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 3));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 2));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 1));
+  }
+
+  {
+    // invalid str - incorrect char 0009 instead of 0008
+    km_core_cu const s[] = {u'a', u'b', 0xFFFF, 0x0009, 0x0001, u'c'};
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 6));      // whole thing
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 5));      // whole thing minus 'c'
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 3));  // just the marker
+    // if we slice the marker, invalid
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 4));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 3));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 2));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 1));
+  }
+
+  {
+    // invalid str - out of range merker 0xDDDD exceeds LDML_MARKER_MAX_INDEX
+    km_core_cu const s[] = {u'a', u'b', 0xFFFF, 0x0008, 0xDDDD, u'c'};
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 6));      // whole thing
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 5));      // whole thing minus 'c'
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 3));  // just the marker
+    // if we slice the marker, invalid
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 4));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 3));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 2));
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s + 2, 1));
+  }
+  {
+    // noncharacter FDD4
+    km_core_cu const s[] = {0xFDD4, 0x0020};
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 1)); // at end of str
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 2)); // not followed by trailing surrogate
+  }
+  {
+    // FFFE nonchar (mismatched BOM)
+    km_core_cu const s[] = {0xFFFE, 0x0020};
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 1)); // at end of str
+    EXPECT_FALSE(COMP_KMXPLUS_STRS::valid_string(s, 2)); // not followed by trailing surrogate
```

Comment:
Some copypasta? With the comments on these tests -- there are no surrogates, and "at end of str" also doesn't fit -- surely all 4 tests are triggered by the first character (U+FDD4 or U+FFFE) and the second character becomes irrelevant? Probably only useful to have 2 tests in that case.

Reply:
Copypasta but checks that it's not only being processed as a 2 char sequence.